lib/multiplexer/multiplexer.go

```diff
@@ -414,75 +433,102 @@ func (m *Mux) detect(conn net.Conn) (*Conn, error) {
 
 		switch proto {
 		case ProtoProxy:
-			if !m.EnableExternalProxyProtocol {
+			newPROXYLine, err := ReadProxyLine(reader)
+			if err != nil {
+				return nil, trace.Wrap(err, invalidProxyLineError)
+			}
+
+			if m.PROXYProtocolMode == PROXYProtocolOff {
 				return nil, trace.BadParameter(proxyProtocolDisabledError)
 			}
-			// We allow only one unsigned proxy line
-			if proxyLine != nil {
-				return nil, trace.BadParameter(duplicateProxyLineError)
+
+			unsignedPROXYLineReceived = true
+
+			if m.PROXYProtocolMode == PROXYProtocolUnspecified {
+				m.Logger.WithFields(log.Fields{
+					"direct_src_addr": conn.RemoteAddr(),
+					"direct_dst_addr": conn.LocalAddr(),
+					"proxy_src_addr:": newPROXYLine.Source.String(),
+					"proxy_dst_addr:": newPROXYLine.Destination.String(),
+				}).Error(unexpectedPROXYLine)
+				newPROXYLine.Source.Port = 0
 			}
-			proxyLine, err = ReadProxyLine(reader)
-			if err != nil {
-				return nil, trace.Wrap(err, invalidProxyLineError)
+
+			if proxyLine != nil {
+				if proxyLine.IsVerified {
+					// signed PROXY line takes precedence
+					continue
+				} else {
+					// We allow only one unsigned PROXY line
+					return nil, trace.BadParameter(duplicateProxyLineError)
+				}
 			}
+
+			proxyLine = newPROXYLine
+
 			// repeat the cycle to detect the protocol
 		case ProtoProxyV2:
-			newProxyLine, err := ReadProxyLineV2(reader)
+			newPROXYLine, err := ReadProxyLineV2(reader)
 			if err != nil {
 				return nil, trace.Wrap(err, invalidProxyV2LineError)
 			}
-			if newProxyLine == nil {
-				continue
+			if newPROXYLine == nil {
+				unsignedPROXYLineReceived = true
```

Comment:
Given this logic, wouldn't multiple PROXYv2 headers containing LOCAL commands be allowed? Should that be the case?

Reply:
Yes, but I would say it doesn't matter since LOCAL command doesn't do anything. We can add checking if `unsignedPROXYLineReceived` is already `true` before setting it and returning error if it is, but I'm not sure if it's worth it to complicate code for that.

Reply:
I might be worth it to at least log and possibly also emit an event when this happens, as I don't think this should happen in a legitimate scenario? Or could it?

Reply:
No, it's not a legitimate scenario. I've refactored it to make sure we don't allow multiple PROXY lines with LOCAL command.